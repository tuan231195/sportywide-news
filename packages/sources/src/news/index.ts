export * from './au-news';
export * from './cnet-news';
export * from './smh-news';
export * from './tech-republic-news';
export * from './factory';
export * from './utils';
