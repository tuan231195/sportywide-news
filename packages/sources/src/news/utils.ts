import axios, { AxiosInstance } from 'axios';
import Cheerio from 'cheerio';
import { JSDOM } from 'jsdom';
import createDOMPurify from 'dompurify';
import adapter from 'axios/lib/adapters/http';
import htmlToText from 'html-to-text';

const window = new JSDOM('').window;
const DOMPurify = createDOMPurify(window);

export function getParsedXml(axios: AxiosInstance, url) {
	return axios.get(url).then(({ data, headers }) => {
		const contentType = headers['content-type'];
		if (
			contentType &&
			[
				'application/xml',
				'application/rss+xml',
				'text/xml; charset=UTF-8',
				'text/xml',
			].includes(contentType)
		) {
			return Cheerio.load(data, {
				xmlMode: true,
			});
		}
		return data;
	});
}

export function resolveLazyLoadedImage($: CheerioStatic, source: Cheerio) {
	source.find('img').each((index, element) => {
		const dataSource = $(element).data('src');
		const currentSource = $(element).attr('src');
		if (dataSource && !currentSource) {
			$(element).attr('src', $(element).data('src'));
		}
	});
	return source;
}

export function getCleanedHTML(source) {
	return DOMPurify.sanitize(source, {
		ALLOW_DATA_ATTR: false,
		FORBID_ATTR: ['style', 'class', 'id', 'width', 'height'],
	});
}

export function getThumbnailUrl(source) {
	const $ = Cheerio.load(source);
	return $('img').eq(0).attr('src');
}

export function getRawText(source) {
	return htmlToText.fromString(source, {
		wordwrap: 130,
		format: {
			text: function (elem) {
				if (elem.parent?.name === 'div') {
					return `${elem.data}\n`;
				} else if (elem.parent?.name === 'span') {
					return `${elem.data} `;
				}
				return elem.data;
			},
		},
	});
}

axios.defaults.adapter = adapter;
export { axios };
