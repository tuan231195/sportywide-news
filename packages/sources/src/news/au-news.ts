import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, resolveLazyLoadedImage } from './utils';
import Cheerio from 'cheerio';
import { getCleanedHTML } from 'src/news/utils';
import { arr } from '@vdtn359/news-utils';

export class AuNews extends DefaultNews {
	static url = 'www.news.com.au';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string): Promise<string> {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		const contentSelectors = ['.story-content', '.story-body'];
		let article = arr.findMap(
			contentSelectors,
			(selector) => $(selector),
			(element) => !!element?.length
		);
		article = resolveLazyLoadedImage($, article);
		const storyContent = Cheerio.html(article);
		return getCleanedHTML(storyContent);
	}
}
