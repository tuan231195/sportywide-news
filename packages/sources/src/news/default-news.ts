import { News } from 'src/news.interface';
import { CATEGORY, getSlug, NewsDto } from '@vdtn359/news-models';
import { getParsedXml } from 'src/news/utils';
import axios, { AxiosInstance } from 'axios';
import { str } from '@vdtn359/news-utils';
import { parse } from 'date-fns';

export abstract class DefaultNews implements News {
	protected axios: AxiosInstance;
	protected constructor(
		protected readonly category: CATEGORY,
		protected rssFeed: string
	) {
		this.axios = axios.create({
			baseURL: rssFeed,
		});
	}

	async extractFeeds(): Promise<NewsDto[]> {
		const $: CheerioStatic = await getParsedXml(this.axios, this.rssFeed);
		return Array.from($('item')).map((node) => {
			const element = $(node);
			const url = element.find('> link').text().trim();
			const title = element.find('> title').text().trim();
			const description = element.find('> description').text().trim();
			const pubDate =
				element.find('pubDate').text() || $('pubDate').text();

			return {
				category: this.category,
				id: str.toGuid(url),
				title,
				description,
				image: this.getImage(element),
				pubDate: parse(
					pubDate,
					'E, dd MMM yyyy HH:mm:ss XX',
					new Date()
				),
				url,
				slug: getSlug(url),
				feed: this.rssFeed.trim(),
			};
		});
	}

	getImage(element: Cheerio): string | undefined {
		const image = element.find('image');
		return image.length ? image.find('url').text().trim() : undefined;
	}
}
