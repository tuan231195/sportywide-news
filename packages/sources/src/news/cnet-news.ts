import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import Cheerio from 'cheerio';
import { getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import { axios } from './utils';

export class CnetNews extends DefaultNews {
	static url = 'www.cnet.com';
	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	getImage(element: Cheerio): string | undefined {
		const image = element.find('media\\:thumbnail');
		return image.length ? image.attr('url')! : undefined;
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		let storyBody = $('#article-body');
		storyBody.find('figure > a > svg').parent().remove();
		const toRemove = [
			'[data-video-playlist]',
			'[section="shortcodeGallery"]',
			'footer',
			'.playerControls',
			'[data-component="lazyloadElement"]',
		];
		toRemove.forEach((section) => {
			storyBody.find(section).remove();
		});
		storyBody = resolveLazyLoadedImage($, storyBody);
		const storyContent = Cheerio.html(storyBody);
		return getCleanedHTML(storyContent);
	}
}
