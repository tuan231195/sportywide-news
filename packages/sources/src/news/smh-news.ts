import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import Cheerio from 'cheerio';
import { arr } from '@vdtn359/news-utils';

export class SmhNews extends DefaultNews {
	static url = 'www.smh.com.au';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		const contentSelectors = [
			'#content > div > div > article > div > div:nth-child(1) > div:nth-child(2)',
			'#content > div > article > section > div',
			'#content > div > article',
		];
		let articleBody = arr.findMap(
			contentSelectors,
			(selector) => $(selector),
			(element) => !!element?.length
		);
		if (!articleBody?.length) {
			return '';
		}
		const toRemove = ['#endOfArticle', '.noPrint', '[preload=metadata]'];
		toRemove.forEach((section) => {
			articleBody.find(section).remove();
		});
		articleBody = resolveLazyLoadedImage($, articleBody);
		const storyContent = Cheerio.html(articleBody);
		return getCleanedHTML(storyContent);
	}
}
