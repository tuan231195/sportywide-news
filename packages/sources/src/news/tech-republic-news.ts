import { DefaultNews } from 'src/news/default-news';
import { CATEGORY } from '@vdtn359/news-models';
import { axios, getCleanedHTML, resolveLazyLoadedImage } from 'src/news/utils';
import Cheerio from 'cheerio';

export class TechRepublicNews extends DefaultNews {
	static url = 'www.techrepublic.com';

	constructor(category: CATEGORY, rssFeed: string) {
		super(category, rssFeed);
	}

	static async extractUrl(url: string) {
		const newsPage = await axios(url).then(({ data }) => data);
		const $ = Cheerio.load(newsPage);
		let articleBody = $('article.article-single > div.content');
		const toRemove = ['[data-component="newsletterSubscription"]'];
		toRemove.forEach((section) => {
			articleBody.find(section).remove();
		});
		articleBody = resolveLazyLoadedImage($, articleBody);
		const storyContent = Cheerio.html(articleBody);
		return getCleanedHTML(storyContent);
	}
}
