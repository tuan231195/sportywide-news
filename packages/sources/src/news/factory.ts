import { CnetNews } from 'src/news/cnet-news';
import { AuNews } from 'src/news/au-news';
import { TechRepublicNews } from 'src/news/tech-republic-news';
import { SmhNews } from 'src/news/smh-news';
import { keyBy } from 'lodash';
import urlParser from 'url';

export const availableSources = [AuNews, CnetNews, TechRepublicNews, SmhNews];
const sourceMap = keyBy(availableSources, 'url');

export function extractUrl(url) {
	const parsedUrl = urlParser.parse(url);
	const host: any = parsedUrl.hostname;
	const source = sourceMap[host];
	if (!source) {
		return;
	}
	return source.extractUrl(url);
}
