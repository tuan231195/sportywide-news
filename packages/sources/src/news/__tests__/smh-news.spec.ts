import { getRawText } from 'src/news/utils';
import { SmhNews } from 'src/news';

describe('#extractNews', () => {
	it('should strip style and css and data attributes', async () => {
		expect(
			getRawText(
				await SmhNews.extractUrl(
					'https://www.smh.com.au/national/queensland/queensland-election-updates-live-latest-on-the-council-election-and-byelections-20200328-p54eul.html'
				)
			)
		).toMatchSnapshot();
	});
});
