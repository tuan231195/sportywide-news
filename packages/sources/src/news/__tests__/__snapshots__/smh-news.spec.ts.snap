// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#extractNews should strip style and css and data attributes 1`] = `
"LATEST UPDATES
 
11.52pm on Mar 28, 2020MAYORAL RACES: WHAT WE KNOW SO FAR
The Electoral Commission of Queensland has stopped the count, with only the unofficial preliminary results publicly available so far.

Here is where we end tonight in key mayoral races.

BRISBANE: (14.3 per cent counted) Lord mayor Adrian Schrinner on 43.47 per cent, Labor candidate Patrick Condren on 30.93 per cent, and Greens candidate Kath Angus on 18.09 per cent.

GOLD COAST: (16.67 per cent counted) Mayor Tom Tate on 54.70 per cent and Mona Hecke on 21.59 per cent.

IPSWICH: (1.34 per cent counted)  Teresa Harding on 40.19 per cent and David Martin on 22.51 per cent.

MORETON BAY: (1.06 per cent counted) Adrian Raedel on 25.88 per cent, Peter Flannery on 24.96 per cent, and Chris Thompson on 20.65 per cent.

LOGAN: Only 0.05 per cent of the count is displaying.

And that's a wrap. The commission will continue the vote count tomorrow and Brisbane Times will have comprehensive coverage of the results.



More 
11.34pm on Mar 28, 2020TECHNICAL ISSUES DELAY RESULTS OF CONTROVERSIAL COUNCIL ELECTIONS
After one of the most unusual council campaigns in Queensland's history - if not Australia's - voters went to the polls today to decide who would lead them through the next four years.

An election official uses sanitiser at the Brisbane City Hall polling booth. [https://static.ffx.io/images/$zoom_0.201%2C$multiply_0.3541%2C$ratio_1.776846%2C$width_1059%2C$x_0%2C$y_84/t_crop_custom/q_86%2Cf_auto/af7b234fa6dccfed13201f857225799033382339]An election official uses sanitiser at the Brisbane City Hall polling booth. Credit: Jono Searle/Getty Images

But by late tonight, for most councils that was still an unknown as the Electoral Commission of Queensland's preliminary vote inched forward slowly and technical issues with the ECQ's website caused headaches.

Only small percentages for most votes were released by late Saturday night. Volunteers dealt not only with counting votes, but managing hygiene measures put in place to protect them during the coronavirus pandemic.

Full story is here.
[https://www.brisbanetimes.com.au/national/queensland/technical-issues-delay-results-of-controversial-council-elections-20200328-p54evl.html]"
`;
